setup.py

```diff
@@ -47,6 +47,20 @@
 
     raise SystemExit("\n".join(msg))
 
+
+# Cocotb requires to link Python dynamic library.
+# On Windows, the Python library is always a dynamic (an import library).
+# Python must be compiled with the --enable-shared flag on Linux
+# or the --enable-framework flag on Mac.
+from distutils import sysconfig
+
+if sys.platform == "darwin" and "--enable-framework" not in sysconfig.get_config_vars().get("CONFIG_ARGS", ""):
+    raise SystemExit("ERROR: You must use a Python compiled with --enable-framework")
+
+if sys.platform == "linux" and "--enable-shared" not in sysconfig.get_config_vars().get("CONFIG_ARGS", ""):
+    raise SystemExit("ERROR: You must use a Python compiled with --enable-shared")
+
```

Comment:
I wonder if we should just dump out as much information about the environment as possible to help the user in case there are multiple installations?
https://docs.python.org/3/library/sysconfig.html#using-sysconfig-as-a-script shows how to do that. The paths are the most important thing, and it says there it uses ``get_path()``, but according to the description, it looks like ``get_paths()`` is actually the function to use? (I cannot try it right now.)

Reply:
At least including `sys.executable` in the message would be a good idea.

I'm also a little worried this might have a false positive. Can we add an override env var to continue with compilation anyway,`COCOTB_OVERRIDE_SHARED_CHECK`?

Reply:
>Can we add an override env var to continue with compilation anyway,COCOTB_OVERRIDE_SHARED_CHECK?

Are you sure we should try to workaround issue that we do know if exist?

Reply:
I suppose we can already do a bugfix release with that change of it did become a problem - so you're right, probably not a good idea after all